### ISSUE TITLE
[webpack 5] moduleIds hashed shouldn't be deprecated for deterministic

### ISSUE THREAD

**Issue by gpoitch** [webpack/webpack#11796]
## Feature request

**What is the expected behavior?**
Webpack 5 presents a warning that moduleIds: hashed is deprecated in favor of deterministic.  These aren't interchangeable when modules are referenced across more than 1 build.

**What is motivation or use case for adding/changing the behavior?**
In a multi-compiler build (e.g. SSR server & client) we collect which modules initiate dynamic imports() in the client bundle and pass that information to the server bundle.  When a request runs on the server bundle, if any of those modules get required(), we know to include that dynamic import's script tag in the server response html.  The module ids have to be hashed in this case so they match across builds.

**How should this be implemented in your opinion?**
Remove the build warning.

**Are you willing to work on this yourself?**
yes

**Comment by evilebottnawi** [comment]
Use should not use `hashed` or `deterministic` in this case

**Comment by gpoitch** [comment]
@evilebottnawi can a function currently be used? Documentation doesn't say that it can. You're right, it would be probably faster to only uniquely identify the modules we need and do the fastest/simplest thing for the rest.

**Comment by evilebottnawi** [comment]
@gpoitch because it is very non standard, I thing you need to implement own plugin, you can look at code of `hashed` plugin

**Comment by gpoitch** [comment]
@evilebottnawi ok, I think that could be doable.  What's the best way to determine if a module is a dynamic import trigger now?  I was checking `module.reasons` but it was removed in webpack 5.  My workaround was calling `compilation.getStats()` which still contains the reasons but I think that wouldn't work during the custom module id'ing phase.

**Comment by evilebottnawi** [comment]
You need graph for this

**Comment by sokra** [comment]
ModuleGraph getIncomingConnections

Yep, best don't use ids for this task. Even with hashes ids there is no guarantee that ids are identical. Use the default ids and store mappings for id to the information you need.

**Comment by gpoitch** [comment]
Thanks for the tips.  I'm realizing the modules ids aren't matching up anymore in webpack 5 prod mode even if I set `hashed`.

I ended up leaving `moduleIds` to its default (unset).

Then in my plugin, used the moduleGraph to find the module for each chunk in the client bundle that triggers a dynamic import and record its dependency `request`.  Then find each module in the server bundle who has matching requests and record their ids.
